inference.py

```diff
@@ -16,133 +16,66 @@
-                            vis_from_batch(params, inputs, outputs, batch_index=0, vis_path=output_path,
-                                           dataset=f'{row_start}_{col_start}_inf', ep_num=index, debug=True)
-
-                        outputs = F.softmax(outputs, dim=1)
-                        output_counts[row_start:row_end, col_start:col_end] += 1
-
-                        # Add inference on sub-image to all completed inferences on previous sub-images.
-                        # TODO: This operation need to be optimized. Using a lot of RAM on large images.
-                        output_probs[:, row_start:row_end, col_start:col_end] += np.squeeze(outputs.cpu().numpy(),
-                                                                                            axis=0)
-
-                        if debug and device.type == 'cuda':
-                            res, mem = gpu_stats(device=device.index)
-                            _tqdm.set_postfix(OrderedDict(gpu_perc=f'{res.gpu} %',
-                                                          gpu_RAM=f'{mem.used / (1024 ** 2):.0f}/{mem.total / (1024 ** 2):.0f} MiB',
-                                                          inp_size=inputs.cpu().numpy().shape,
-                                                          out_size=outputs.cpu().numpy().shape,
-                                                          overlay=overlay))
-
-            # Divide array according to output counts. Manages overlap and returns a softmax array as if only one forward pass had been done.
-            output_mask_raw = np.divide(output_probs, np.maximum(output_counts, 1))  # , 1 is added to overwrite 0 values.
-
-            # Resize the output array to the size of the input image and write it
-            output_mask_raw_cropped = np.moveaxis(output_mask_raw, 0, -1)
-            output_mask_raw_cropped = output_mask_raw_cropped[overlay:(h + overlay), overlay:(w + overlay), :]
-
-            return output_mask_raw_cropped
-    else:
-        raise IOError(f"Error classifying image : Image shape of {len(nd_array.shape)} is not recognized")
-
+    img_array, input_image, dataset_nodata = image_reader_as_array(input_image=raster,
+                                                                   clip_gpkg=clip_gpkg)
+    metadata = add_metadata_from_raster_to_sample(img_array, 
+                                                    input_image, 
+                                                    meta_map=None, 
+                                                    raster_info=None)
+    h, w, num_bands = img_array.shape
+    h_ = sample_size * math.ceil(h / sample_size)
+    w_ = sample_size * math.ceil(w / sample_size)
+    pred_img = np.empty((h_, w_), dtype=np.uint8)
+    for row in tqdm(range(0, h, sample_size), position=1, leave=False, desc='Inferring rows'):
+        with tqdm(range(0, w, sample_size), position=2, leave=False, desc='Inferring columns') as _tqdm:
+            for column in _tqdm:
```

Comment:
Since you're not using the overlap parameter, have you noticed inconsistencies in the resulting raster, like those showed in https://github.com/NRCan/geo-deep-learning/issues/11

Reply:
Hmmn, I had no idea the overlap solves that problem. I wanted to implement this https://github.com/Vooban/Smoothly-Blend-Image-Patches to solve that problem. If the overlap solves that problem I would be happy to retain it.

Reply:
Well, the overlap reduced the impact of this problem but did not solved it completely.  

Reading what's in the link you provided, I really prefer your approach. The first advantage I see is that it allows for data parallelization, which @remtav will certainly be glad to see!

Reply:
Parallelization would be cool. However I don't see the interest in using segmentation without smoothing. We're keeping only as a basice version of inference?

Reply:
Yes, plus the smoothing can be fine-tuned better. Nice to have a base version for comparison. it can be deprecated in the future.

Reply:
Ok. My only (final!) concern is that infering with an overlap (commonly 25%) in the previous version of inference.py gave much better results than no overlap (i.e. what the base version would output). 
Although it's great in terms of memory efficiency and necessary computing, I'm not convinced the smoothing version will give an output just as good as inference with 25% overlap. On the demo you gave, I saw quite a few edges where smoothing introduced other values (ex.: vegetation in hydro or the opposite). I'd have to see it again, but anyways my concern is smoothing may not perform as well as previous inference with overlap.

Reply:
I think the base version can be included with overlap. No problems. I can create a new ticket for that.